flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java

```diff
@@ -751,6 +753,34 @@ public void dropPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSp
 		}
 	}
 
+	public List<CatalogPartitionSpec> listPartitionsByFilter(ObjectPath tablePath, List<Expression> expressions)
+			throws TableNotExistException, TableNotPartitionedException, CatalogException {
+		Table hiveTable = getHiveTable(tablePath);
+		ensurePartitionedTable(tablePath, hiveTable);
+		List<String> partColNames = getFieldNames(hiveTable.getPartitionKeys());
+		String filter = HiveTableUtil.makePartitionFilter(partColNames, expressions);
+		if (filter == null) {
+			throw new UnsupportedOperationException(
+					"HiveCatalog is unable to handle the partition filter expressions: " + expressions);
+		}
+		try {
+			PartitionSpecProxy.PartitionIterator partitions = client.listPartitionSpecsByFilter(
+					tablePath.getDatabaseName(), tablePath.getObjectName(), filter, (short) -1).getPartitionIterator();
+			List<CatalogPartitionSpec> res = new ArrayList<>();
+			while (partitions.hasNext()) {
+				Partition partition = partitions.next();
+				Map<String, String> spec = new HashMap<>();
+				for (int i = 0; i < partColNames.size(); i++) {
+					spec.put(partColNames.get(i), partition.getValues().get(i));
+				}
+				res.add(new CatalogPartitionSpec(spec));
+			}
+			return res;
+		} catch (TException e) {
+			throw new UnsupportedOperationException("Failed to list partition by filter from HMS", e);
```

Comment:
why UnsupportedOperationException?

Reply:
It's possible the underlying DB doesn't support the filter, in which case we would want to let planner fall back to list all partitions and filter by itself.

Reply:
who would catch the exception, and do something about it?

I feel it better be a declared exception to notify callers they need to watch out for it, like a new IllegalFilterException? UnsupportedOperationException is for operation that is universally unsupported, but the case here is about input args where some **args** (filters) are supported while some are not.

Besides, add the filter to error msg?

Reply:
The exception is defined in the Catalog API contract:
```
	 * <p>If catalog does not support this interface at present, throw an {@link UnsupportedOperationException}
	 * directly. If the catalog does not have a valid filter, throw the {@link UnsupportedOperationException}
	 * directly. Planner will fallback to get all partitions and filter by itself.
```
I'll add the expressions to error msg.